File: src/alpha/Input.js
// Input Version 1.2.130825
// usage:
// local input = Input:New(glwidget)
// input:SetMouseSensitivityX( .05 ) // defaults to .05
// input:SetMouseSensitivityY( .05 ) // defaults to .05

// inside of a timing.every function
// Camera:MoveForward( input:W() * elapsed );
// Camera:YawLeft( input:LeftMouseButton() * input:MouseLeft() )
// some non-obvious buttons are:
// LeftMouseButton, RightMouseButton, MiddleMouseButton, SPACE, RETURN, SHIFT
// MouseUp, MouseDown, MouseLeft, MouseRight

// for a simple if statement do:
//       if input:Q() > 0 then
// do stuff because Q is down
//       end

// MouseWheelUp()
// returns 1 or more if you have scrolled up recently
// MouseWheelDegreesUp()
// returns the number of degrees the wheel has scrolled recently

// add this to your code to make a command only work once per button push
/* eslint-disable require-jsdoc, new-cap */
/*
        if elapsed == 0 then
                done = false;
                return

        end
        if done then return end;
        done = true;
*/

export function AlphaInput(surface, camera) {
  this.SetMouseSensitivityX(0.005);
  this.SetMouseSensitivityY(0.005);

  this.surface = surface;
  this.camera = camera;
  this.startX = 0;
  this.endX = 0;
  this.startY = 0;
  this.endY = 0;
  this.mouseWheelUp = 0;
  this.mouseWheelDown = 0;
  this.grabbed = null;
}

AlphaInput.prototype.onKeyup = function(event) {
  this[event.key.toLowerCase()] = null;
  return true;
};

AlphaInput.prototype.onKeydown = function(event) {
  if (this.onKeyDown(event.key)) {
    return;
  }
  if (event.ctrlKey || event.altKey || event.metaKey) {
    return;
  }
  this[event.key.toLowerCase()] = 1;
  return true;
};

export function alphaGetButtonName(buttonIndex) {
  switch (event.button) {
    case 0:
      return 'LeftMouseButton';
    case 2:
      return 'RightMouseButton';
    case 1:
      return 'MiddleMouseButton';
  }
  return null;
}

AlphaInput.prototype.onMousedown = function(event) {
  let button;
  let x;
  let y;
  button = alpha_GetButtonName(event.button);
  x = event.x;
  y = event.y;
  this[button] = 1;

  // reset for a new drag
  this.startX = x;
  this.startY = y;
  this.endX = x;
  this.endY = y;
  return true;
};

AlphaInput.prototype.onMouseup = function(event) {
  let button;
  let x;
  let y;
  button = alpha_GetButtonName(event.button);
  x = event.clientX;
  y = event.clientY;
  this[button] = null;

  // new end point;
  this.endX = x;
  this.endY = y;
  return true;
};

AlphaInput.prototype.onMousemove = function(event) {
  let x;
  let y;
  x = event.x;
  y = event.y;
  this.endX = x;
  this.endY = y;
  return true;
};

AlphaInput.prototype.onWheel = function(event) {
  const wheel = event.spinY;
  if (wheel > 0) {
    this.mouseWheelUp = this.mouseWheelUp + wheel;
  } else {
    // keeping it positive!
    this.mouseWheelDown = this.mouseWheelDown - wheel;
  }
  return true;
};

AlphaInput.prototype.onKeyDown = function(...args) {
  if (this._keyDownListener) {
    return this._keyDownListener.apply(this._keyDownThisObject, args);
  }
  return false;
};

AlphaInput.prototype.SetOnKeyDown = function(listener, thisObject) {
  this._keyDownListener = listener;
  this._keyDownThisObject = thisObject;
};

AlphaInput.prototype.Get = function(key) {
  return this[key] ? 1 : 0;
};

AlphaInput.prototype.SetMouseSensitivityX = function(sensitivity) {
  this.mouseSensitivityX = sensitivity;
};

AlphaInput.prototype.GetMouseSensitivityX = function() {
  return this.mouseSensitivityX;
};

AlphaInput.prototype.SetMouseSensitivityY = function(sensitivity) {
  this.mouseSensitivityY = sensitivity;
};

AlphaInput.prototype.GetMouseSensitivityY = function() {
  return this.mouseSensitivityY;
};

// quick set both of them
AlphaInput.prototype.SetMouseSensitivity = function(sensitivity) {
  this.SetMouseSensitivityX(sensitivity);
  this.SetMouseSensitivityY(sensitivity);
};

AlphaInput.prototype.MouseLeft = function() {
  if (this.endX < this.startX) {
    const change = this.startX - this.endX;
    // console.log("mouse has moved right " + change);
    return change * this.GetMouseSensitivityX();
  }

  return 0;
};

AlphaInput.prototype.MouseRight = function() {
  if (this.endX > this.startX) {
    const change = this.endX - this.startX;
    // console.log("mouse has moved left " + change);
    return change * this.GetMouseSensitivityX();
  }

  return 0;
};

AlphaInput.prototype.MouseUp = function() {
  if (this.endY > this.startY) {
    const change = this.endY - this.startY;
    // console.log("mouse has moved down " + change);
    return change * this.GetMouseSensitivityY();
  }

  return 0;
};

AlphaInput.prototype.MouseDown = function() {
  if (this.endY < this.startY) {
    const change = this.endY - this.startY;
    // console.log("mouse has moved up " + change);
    return change * this.GetMouseSensitivityY();
  }

  return 0;
};

// mouse wheel data is stored in 1/8 of a degree
// this returns how many ticks of a mousewheel of standard resolution
// has been seen before an Input:Update()
AlphaInput.prototype.MouseWheelUp = function() {
  return this.mouseWheelUp / 120;
};

AlphaInput.prototype.MouseWheelDown = function() {
  return this.mouseWheelDown / 120;
};

AlphaInput.prototype.MouseWheelDegreesUp = function() {
  return this.mouseWheelUp / 8;
};

AlphaInput.prototype.MouseWheelDegreesDown = function() {
  return this.mouseWheelDown / 8;
};

/*
 * Sets the start to the end, and clears mousewheel totals.
 */
AlphaInput.prototype.Update = function(elapsed) {
  // console.log("Updating with elapsed: " + elapsed);
  if (this.Get('Shift') > 0) {
    elapsed = elapsed * 10;
  }

  if (this.Get('Shift') > 0) {
    elapsed = elapsed / 10;
  }

  // console.log("LeftMouseButton: " + this.Get("LeftMouseButton"));
  // console.log("MouseLeft: " + this.MouseLeft() * elapsed);
  // console.log("MouseLeft: " +
  //   (this.Get("LeftMouseButton") * this.MouseLeft() * elapsed));
  // console.log("LeftMouse: " + this.Get("LeftMouseButton"));
  // console.log("TurnLeft: " + this.MouseLeft() * elapsed);
  this.camera
      .GetParent()
      .TurnLeft(this.Get('LeftMouseButton') * this.MouseLeft() * elapsed);
  this.camera
      .GetParent()
      .TurnRight(this.Get('LeftMouseButton') * this.MouseRight() * elapsed);
  this.camera
      .GetParent()
      .PitchUp(-this.Get('LeftMouseButton') * this.MouseUp() * elapsed);
  this.camera
      .GetParent()
      .PitchDown(this.Get('LeftMouseButton') * this.MouseDown() * elapsed);
  this.camera.MoveForward(this.MouseWheelDegreesUp() * elapsed);
  this.camera.MoveBackward(this.MouseWheelDegreesDown() * elapsed);
  // this.camera.ZoomIn(this.Get("y"), elapsed);
  // this.camera.ZoomOut(this.Get("h"), elapsed);

  this.camera.GetParent().MoveForward(100 * this.Get('t') * elapsed);
  this.camera.GetParent().MoveBackward(100 * this.Get('g') * elapsed);
  this.camera.GetParent().MoveLeft(100 * this.Get('f') * elapsed);
  this.camera.GetParent().MoveRight(100 * this.Get('h') * elapsed);

  this.camera.GetParent().MoveForward(this.Get('w') * elapsed);
  this.camera.GetParent().MoveBackward(this.Get('s') * elapsed);
  this.camera.GetParent().MoveLeft(this.Get('a') * elapsed);
  this.camera.GetParent().MoveRight(this.Get('d') * elapsed);
  this.camera.GetParent().MoveUp(this.Get(' ') * elapsed);
  this.camera.GetParent().MoveDown(this.Get('Shift') * elapsed);

  this.camera.GetParent().YawLeft(this.Get('j') * elapsed);
  this.camera.GetParent().YawRight(this.Get('l') * elapsed);
  this.camera.GetParent().PitchUp(this.Get('k') * elapsed);
  this.camera.GetParent().PitchDown(this.Get('i') * elapsed);
  this.camera.GetParent().RollLeft(this.Get('u') * elapsed);
  this.camera.GetParent().RollRight(this.Get('o') * elapsed);

  if (this.Get('RightMouseButton') > 0) {
    if (!this._done) {
      this.camera.AlignParentToMy(false, true);
      this._done = true;
    }
  } else {
    this._done = false;
  }
  this.startX = this.endX;
  this.startY = this.endY;
  this.mouseWheelUp = 0;
  this.mouseWheelDown = 0;
};

File: src/alpha/Input.c
#include "Input.h"
#include <stdio.h>
#include "Maths.h"
#include "../graph/log.h"

alpha_Input* alpha_Input_new(apr_pool_t* pool, alpha_Camera* camera)
{
    alpha_Input* input;
    if(pool) {
        input = apr_palloc(pool, sizeof(*input));
    }
    else {
        input = malloc(sizeof(*input));
    }
    input->pool = pool;

    alpha_Input_SetMouseSensitivityX(input, .005);
    alpha_Input_SetMouseSensitivityY(input, .005);

    input->keyPress = apr_hash_make(pool);
    input->camera = camera;
    input->startX = 0;
    input->endX = 0;
    input->startY = 0;
    input->endY = 0;
    input->mouseWheelUp = 0;
    input->mouseWheelDown = 0;
    input->grabbed = 0;

    return input;
};

void alpha_Input_keyup(alpha_Input* input, const char* key)
{
    apr_hash_set(input->keyPress, key, APR_HASH_KEY_STRING, (const void*)0);
};

void alpha_Input_keydown(alpha_Input* input, const char* key, int ctrlKey, int altKey, int metaKey)
{
    if(ctrlKey || altKey || metaKey) {
        return;
    }
    //parsegraph_log("keydown %s\n", key);
    apr_hash_set(input->keyPress, key, APR_HASH_KEY_STRING, (const void*)1);
}

void alpha_Input_mousedown(alpha_Input* input, const char* button, int x, int y)
{
    apr_hash_set(input->keyPress, button, APR_HASH_KEY_STRING, (const void*)1);

    // reset for a new drag
    input->startX = x;
    input->startY = y;
    input->endX = x;
    input->endY = y;
}

void alpha_Input_mouseup(alpha_Input* input, const char* button, int x, int y)
{
    apr_hash_set(input->keyPress, button, APR_HASH_KEY_STRING, (const void*)0);

    // new end point;
    input->endX = x;
    input->endY = y;
}

void alpha_Input_mousemove(alpha_Input* input, int x, int y)
{
    input->endX = x;
    input->endY = y;
};

void alpha_Input_onWheel(alpha_Input* input, int wheel)
{
    if(wheel > 0) {
        input->mouseWheelUp = input->mouseWheelUp + wheel;
    }
    else if(wheel < 0) {
        // keeping it positive!
        input->mouseWheelDown = input->mouseWheelDown - wheel;
    }
    else {
        // I have no idea how I got here
    }
}

int alpha_Input_Get(alpha_Input* input, const char* key)
{
    void* val = apr_hash_get(input->keyPress, key, APR_HASH_KEY_STRING);
    return val != 0;
};

void alpha_Input_SetMouseSensitivityX(alpha_Input* input, float sensitivity)
{
    input->mouseSensitivityX = sensitivity;
}

float alpha_Input_GetMouseSensitivityX(alpha_Input* input)
{
    return input->mouseSensitivityX;
}

void alpha_Input_SetMouseSensitivityY(alpha_Input* input, float sensitivity)
{
    input->mouseSensitivityY = sensitivity;
}

float alpha_Input_GetMouseSensitivityY(alpha_Input* input)
{
    return input->mouseSensitivityY;
}

// quick set both of them
void alpha_Input_SetMouseSensitivity(alpha_Input* input, float sensitivity)
{
    alpha_Input_SetMouseSensitivityX(input, sensitivity);
    alpha_Input_SetMouseSensitivityY(input, sensitivity);
}

float alpha_Input_MouseLeft(alpha_Input* input)
{
    if(input->endX < input->startX) {
        float change = input->startX - input->endX;
        //parsegraph_log("mouse has moved right %d", change);
        return change * alpha_Input_GetMouseSensitivityX(input);
    }

    return 0;
};

float alpha_Input_MouseRight(alpha_Input* input)
{
    if(input->endX > input->startX) {
        float change = input->endX - input->startX;
        //fprintf(stderr "mouse has moved left %d", change);
        return change * alpha_Input_GetMouseSensitivityX(input);
    }

    return 0;
};

float alpha_Input_MouseUp(alpha_Input* input)
{
    if(input->endY > input->startY) {
        float change = input->endY - input->startY;
        //fprintf("mouse has moved down %d", change);
        return change * alpha_Input_GetMouseSensitivityY(input);
    }

    return 0;
};

float alpha_Input_MouseDown(alpha_Input* input)
{
    if(input->endY < input->startY) {
        float change = input->endY - input->startY;
        //parsegraph_log("mouse has moved up %d", change);
        return change * alpha_Input_GetMouseSensitivityY(input);
    }

    return 0;
};

// mouse wheel data is stored in 1/8 of a degree
// this returns how many ticks of a mousewheel of standard resolution
// has been seen before an Input:Update()
float alpha_Input_MouseWheelUp(alpha_Input* input, float degrees)
{
    return input->mouseWheelUp / 120;
};

float alpha_Input_MouseWheelDown(alpha_Input* input)
{
    return input->mouseWheelDown / 120;
};

float alpha_Input_MouseWheelDegreesUp(alpha_Input* input)
{
    return input->mouseWheelUp / 8;
};

float alpha_Input_MouseWheelDegreesDown(alpha_Input* input)
{
    return input->mouseWheelDown / 8;
};

/**
 * Sets the start to the end, and clears mousewheel totals.
 */
void alpha_Input_Update(alpha_Input* input, float elapsed)
{
    if(alpha_Input_Get(input, "Shift") > 0) {
        elapsed = elapsed * 10;
    }

    if(alpha_Input_Get(input, "Shift") > 0) {
        elapsed = elapsed / 10;
    }

    //console.log("LeftMouseButton: " + alpha_Input_Get(input, "LeftMouseButton"));
    //console.log("MouseLeft: " + alpha_Input_MouseLeft(input) * elapsed);
    //console.log("MouseLeft: " + (alpha_Input_Get(input, "LeftMouseButton") * alpha_Input_MouseLeft(input) * elapsed));
    //console.log("LeftMouse: " + alpha_Input_Get(input, "LeftMouseButton"));
    //console.log("TurnLeft: " + alpha_Input_MouseLeft(input) * elapsed);
    alpha_Camera_TurnLeft(input->camera,
        alpha_Input_Get(input, "LeftMouseButton") * alpha_Input_MouseLeft(input) * elapsed
    );
    alpha_Camera_TurnRight(input->camera,
        alpha_Input_Get(input, "LeftMouseButton") * alpha_Input_MouseRight(input) * elapsed
    );
    alpha_Camera_PitchUp(input->camera,
        alpha_Input_Get(input, "LeftMouseButton") * alpha_Input_MouseUp(input) * elapsed
    );
    alpha_Camera_PitchDown(input->camera,
        alpha_Input_Get(input, "LeftMouseButton") * alpha_Input_MouseDown(input) * elapsed
    );
    alpha_Camera_MoveForward(input->camera, alpha_Input_MouseWheelDegreesUp(input) * elapsed);
    alpha_Camera_MoveBackward(input->camera, alpha_Input_MouseWheelDegreesDown(input) * elapsed);
    //alpha_Camera_ZoomIn(input->camera, alpha_Input_Get(input, "y"), elapsed);
    //alpha_Camera_ZoomOut(input->camera, alpha_Input_Get(input, "h"), elapsed);

    if(!alpha_Camera_GetParentIsPhysical(input->camera)) {
        alpha_Camera_MoveForward(alpha_Camera_GetParent(input->camera), alpha_Input_Get(input, "w") * elapsed );
        alpha_Camera_MoveBackward(alpha_Camera_GetParent(input->camera), alpha_Input_Get(input, "s") * elapsed );
    }
    else {
        alpha_Physical* parent = alpha_Camera_GetParent(input->camera);
        //parsegraph_log("W state=%d", alpha_Input_Get(input, "w"));
        alpha_Physical_MoveForward(parent, 100*alpha_Input_Get(input, "t") * elapsed );
        alpha_Physical_MoveBackward(parent, 100*alpha_Input_Get(input, "g") * elapsed );
        alpha_Physical_MoveLeft(parent, 100*alpha_Input_Get(input, "f") * elapsed );
        alpha_Physical_MoveRight(parent, 100*alpha_Input_Get(input, "h") * elapsed );
        alpha_Physical_MoveForward(parent, alpha_Input_Get(input, "w") * elapsed );
        alpha_Physical_MoveBackward(parent, alpha_Input_Get(input, "s") * elapsed );
        alpha_Physical_MoveLeft(parent, alpha_Input_Get(input, "a") * elapsed );
        alpha_Physical_MoveRight(parent, alpha_Input_Get(input, "d") * elapsed );
        alpha_Physical_MoveUp(parent, alpha_Input_Get(input, " ") * elapsed );
        alpha_Physical_MoveDown(parent, alpha_Input_Get(input, "Shift") * elapsed );
        alpha_Physical_YawLeft(parent, alpha_Input_Get(input, "j") * elapsed );
        alpha_Physical_YawRight(parent, alpha_Input_Get(input, "l") * elapsed );
        alpha_Physical_PitchUp(parent, alpha_Input_Get(input, "k") * elapsed );
        alpha_Physical_PitchDown(parent, alpha_Input_Get(input, "i") * elapsed );
        alpha_Physical_RollLeft(parent, alpha_Input_Get(input, "u") * elapsed );
        alpha_Physical_RollRight(parent, alpha_Input_Get(input, "o") * elapsed );
    }

    if(alpha_Input_Get(input, "RightMouseButton") > 0) {
        if(!input->_done) {
            alpha_Camera_AlignParentToMy(input->camera, 0, 1);
            input->_done = 1;
        }
    }
    else {
        input->_done = 0;
    }
    input->startX = input->endX;
    input->startY = input->endY;
    input->mouseWheelUp = 0;
    input->mouseWheelDown = 0;
};
